File: crypto/core/aes_utils.py
"""
AES Utilities for Hybrid Cryptography System
Handles AES file encryption/decryption, key generation, and file information
"""

import os
import hashlib
import mimetypes
from pathlib import Path
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

def generate_aes_key(key_size=32):
    """
    Generate a random AES key
    
    Args:
        key_size (int): Key size in bytes (16=128-bit, 24=192-bit, 32=256-bit)
        
    Returns:
        bytes: Random AES key
    """
    if key_size not in [16, 24, 32]:
        print(f"[WARNING] Invalid key size {key_size}, using 32 bytes (AES-256)")
        key_size = 32
    
    key = os.urandom(key_size)
    print(f"[SUCCESS] Generated AES-{key_size*8} key")
    return key

def get_file_info(file_path):
    """
    Get comprehensive file information
    
    Args:
        file_path (str): Path to file
        
    Returns:
        dict: File information including size, type, hash
    """
    file_path = Path(file_path)
    
    if not file_path.exists():
        raise FileNotFoundError(f"File not found: {file_path}")
    
    # Get file stats
    stat_info = file_path.stat()
    
    # Determine MIME type
    mime_type, encoding = mimetypes.guess_type(str(file_path))
    if mime_type is None:
        mime_type = "application/octet-stream"
    
    # Calculate SHA-256 hash
    sha256_hash = hashlib.sha256()
    with open(file_path, 'rb') as f:
        while chunk := f.read(8192):
            sha256_hash.update(chunk)
    
    file_info = {
        "name": file_path.name,
        "path": str(file_path),
        "extension": file_path.suffix.lower(),
        "size_bytes": stat_info.st_size,
        "size_mb": round(stat_info.st_size / (1024 * 1024), 3),
        "mime_type": mime_type,
        "encoding": encoding,
        "sha256_hash": sha256_hash.hexdigest()
    }
    
    print(f"[INFO] File analyzed: {file_info['name']} ({file_info['size_mb']} MB, {mime_type})")
    return file_info

def encrypt_file(input_file_path, output_file_path, aes_key):
    """
    Encrypt a file using AES-GCM
    
    Args:
        input_file_path (str): Path to input file
        output_file_path (str): Path to output encrypted file
        aes_key (bytes): AES encryption key
        
    Returns:
        tuple: (success, iv, tag, file_info) or (False, None, None, None)
    """
    try:
        # Get file info first
        file_info = get_file_info(input_file_path)
        
        # Generate random IV (96-bit for GCM)
        iv = os.urandom(12)
        
        # Create cipher
        cipher = Cipher(algorithms.AES(aes_key), modes.GCM(iv))
        encryptor = cipher.encryptor()
        
        # Ensure output directory exists
        os.makedirs(os.path.dirname(output_file_path), exist_ok=True)
        
        # Encrypt file
        with open(input_file_path, 'rb') as infile, open(output_file_path, 'wb') as outfile:
            while chunk := infile.read(8192):
                encrypted_chunk = encryptor.update(chunk)
                outfile.write(encrypted_chunk)
        
        # Finalize encryption and get authentication tag
        encryptor.finalize()
        tag = encryptor.tag
        
        print(f"[SUCCESS] File encrypted: {os.path.basename(output_file_path)}")
        return True, iv, tag, file_info
        
    except Exception as e:
        print(f"[ERROR] Encryption failed: {e}")
        return False, None, None, None

def decrypt_file(encrypted_file_path, output_file_path, aes_key, iv, tag):
    """
    Decrypt a file using AES-GCM
    
    Args:
        encrypted_file_path (str): Path to encrypted file
        output_file_path (str): Path to output decrypted file
        aes_key (bytes): AES decryption key
        iv (bytes): Initialization vector
        tag (bytes): Authentication tag
        
    Returns:
        bool: True if successful, False otherwise
    """
    try:
        # Create cipher for decryption
        cipher = Cipher(algorithms.AES(aes_key), modes.GCM(iv, tag))
        decryptor = cipher.decryptor()
        
        # Ensure output directory exists
        os.makedirs(os.path.dirname(output_file_path), exist_ok=True)
        
        # Decrypt file
        with open(encrypted_file_path, 'rb') as infile, open(output_file_path, 'wb') as outfile:
            while chunk := infile.read(8192):
                decrypted_chunk = decryptor.update(chunk)
                outfile.write(decrypted_chunk)
        
        # Finalize decryption (this verifies the authentication tag)
        decryptor.finalize()
        
        print(f"[SUCCESS] File decrypted: {os.path.basename(output_file_path)}")
        return True
        
    except Exception as e:
        print(f"[ERROR] Decryption failed: {e}")
        return False

def verify_file_integrity(file_path, expected_hash):
    """
    Verify file integrity using SHA-256 hash
    
    Args:
        file_path (str): Path to file to verify
        expected_hash (str): Expected SHA-256 hash
        
    Returns:
        bool: True if hash matches, False otherwise
    """
    try:
        sha256_hash = hashlib.sha256()
        with open(file_path, 'rb') as f:
            while chunk := f.read(8192):
                sha256_hash.update(chunk)
        
        actual_hash = sha256_hash.hexdigest()
        
        if actual_hash == expected_hash:
            print("[SUCCESS] File integrity verified ✓")
            return True
        else:
            print("[WARNING] File integrity check failed ✗")
            print(f"Expected: {expected_hash}")
            print(f"Actual:   {actual_hash}")
            return False
            
    except Exception as e:
        print(f"[ERROR] Integrity verification failed: {e}")
        return False

def get_recommended_key_size(file_size_mb, file_type):
    """
    Get recommended AES key size based on file characteristics
    
    Args:
        file_size_mb (float): File size in megabytes
        file_type (str): MIME type of file
        
    Returns:
        int: Recommended key size in bytes
    """
    # Always use AES-256 (32 bytes) for maximum security
    # This is the current best practice
    return 32

if __name__ == "__main__":
    # Test AES functionality
    print("🧪 Testing AES Utilities")
    print("=" * 30)
    
    # Create test file
    test_file = "test_file.txt"
    test_content = "Hello, World! This is a test file for AES encryption.\nMultiple lines of text to test encryption."
    
    with open(test_file, 'w') as f:
        f.write(test_content)
    
    # Generate AES key
    aes_key = generate_aes_key()
    
    # Test encryption
    encrypted_file = "test_file.enc"
    success, iv, tag, file_info = encrypt_file(test_file, encrypted_file, aes_key)
    
    if success:
        print(f"Original hash: {file_info['sha256_hash']}")
        
        # Test decryption
        decrypted_file = "test_file_decrypted.txt"
        decrypt_success = decrypt_file(encrypted_file, decrypted_file, aes_key, iv, tag)
        
        if decrypt_success:
            # Verify integrity
            if verify_file_integrity(decrypted_file, file_info['sha256_hash']):
                print("✅ AES encryption/decryption test PASSED!")
            else:
                print("❌ Integrity verification FAILED!")
        else:
            print("❌ Decryption FAILED!")
    else:
        print("❌ Encryption FAILED!")
    
    # Cleanup
    for f in [test_file, encrypted_file, decrypted_file]:
        if os.path.exists(f):
            os.remove(f)

File: backend/src/main/java/com/sentra/crypto/core/AESUtils.java
package com.sentra.crypto.core;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.HashMap;
import java.util.Map;

import javax.crypto.Cipher;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import com.sentra.backend.runtime.Log;

public class AESUtils {
    private static final int DEFAULT_KEY_SIZE_BYTES = 32;
    private static final int GCM_IV_LENGTH = 12;
    private static final int GCM_TAG_LENGTH_BITS = 128;

    public static byte[] generateAesKey() {
        return generateAesKey(DEFAULT_KEY_SIZE_BYTES);
    }

    public static byte[] generateAesKey(int keySizeBytes) {
        if (keySizeBytes != 16 && keySizeBytes != 24 && keySizeBytes != 32) {
            Log.warn("CRYPTO", "aes_key", "invalid size, using AES-256");
            keySizeBytes = 32;
        }
        byte[] key = new byte[keySizeBytes];
        new SecureRandom().nextBytes(key);
        Log.debug("CRYPTO", "aes_key", "generated size=" + (keySizeBytes * 8));
        return key;
    }

    public static Map<String, Object> getFileInfo(Path filePath) throws IOException {
        if (!Files.exists(filePath)) {
            throw new IOException("File not found: " + filePath);
        }

        long sizeBytes = Files.size(filePath);
        double sizeMb = Math.round((sizeBytes / (1024.0 * 1024.0)) * 1000.0) / 1000.0;

        String mimeType = Files.probeContentType(filePath);
        if (mimeType == null) {
            mimeType = "application/octet-stream";
        }

        String encoding = null;
        String sha256Hash;
        try {
            sha256Hash = sha256Hex(filePath);
        } catch (Exception e) {
            throw new IOException(e);
        }

        Map<String, Object> info = new HashMap<>();
        info.put("name", filePath.getFileName().toString());
        info.put("path", filePath.toString());
        String extension = "";
        String name = filePath.getFileName().toString();
        if (name.contains(".")) {
            extension = name.substring(name.lastIndexOf('.')).toLowerCase();
        }
        info.put("extension", extension);
        info.put("size_bytes", sizeBytes);
        info.put("size_mb", sizeMb);
        info.put("mime_type", mimeType);
        info.put("encoding", encoding);
        info.put("sha256_hash", sha256Hash);

        Log.debug("CRYPTO", "file_info", "name=" + info.get("name") + " size_mb=" + info.get("size_mb"));
        return info;
    }

    public static EncryptionResult encryptFile(Path inputFilePath, Path outputFilePath, byte[] aesKey) {
        try {
            Map<String, Object> fileInfo = getFileInfo(inputFilePath);

            byte[] iv = new byte[GCM_IV_LENGTH];
            new SecureRandom().nextBytes(iv);

            byte[] plaintext = Files.readAllBytes(inputFilePath);
            byte[] combined = encryptAesGcmToCombined(plaintext, aesKey, iv);

            int tagLength = 16;
            int ctLength = combined.length - tagLength;
            byte[] ciphertext = new byte[ctLength];
            byte[] tagBytes = new byte[tagLength];
            System.arraycopy(combined, 0, ciphertext, 0, ctLength);
            System.arraycopy(combined, ctLength, tagBytes, 0, tagLength);

            Files.createDirectories(outputFilePath.getParent());
            try (OutputStream out = Files.newOutputStream(outputFilePath)) {
                out.write(ciphertext);
            }

            Log.debug("CRYPTO", "encrypt", "file=" + outputFilePath.getFileName());
            return new EncryptionResult(true, iv, tagBytes, fileInfo);

        } catch (Exception e) {
            Log.error("CRYPTO", "encrypt", "failed");
            return new EncryptionResult(false, null, null, null);
        }
    }

    public static boolean decryptFile(Path encryptedFilePath, Path outputFilePath, byte[] aesKey, byte[] iv, byte[] tag) {
        try {
            Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
            SecretKeySpec keySpec = new SecretKeySpec(aesKey, "AES");
            GCMParameterSpec gcmSpec = new GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv);
            cipher.init(Cipher.DECRYPT_MODE, keySpec, gcmSpec);

            Files.createDirectories(outputFilePath.getParent());

            byte[] ciphertext = Files.readAllBytes(encryptedFilePath);
            byte[] combined = new byte[ciphertext.length + tag.length];
            System.arraycopy(ciphertext, 0, combined, 0, ciphertext.length);
            System.arraycopy(tag, 0, combined, ciphertext.length, tag.length);

            byte[] plaintext = cipher.doFinal(combined);
            Files.write(outputFilePath, plaintext);

            Log.debug("CRYPTO", "decrypt", "file=" + outputFilePath.getFileName());
            return true;

        } catch (Exception e) {
            Log.error("CRYPTO", "decrypt", "failed");
            return false;
        }
    }

    public static boolean verifyFileIntegrity(Path filePath, String expectedHash) {
        try {
            String actualHash = sha256Hex(filePath);
            if (actualHash.equals(expectedHash)) {
                Log.debug("CRYPTO", "integrity", "verified");
                return true;
            }
            Log.warn("CRYPTO", "integrity", "mismatch");
            return false;
        } catch (Exception e) {
            Log.error("CRYPTO", "integrity", "failed");
            return false;
        }
    }

    public static int getRecommendedKeySize(double fileSizeMb, String fileType) {
        return 32;
    }

    private static String sha256Hex(Path filePath) throws Exception {
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        try (InputStream in = Files.newInputStream(filePath)) {
            byte[] buffer = new byte[8192];
            int read;
            while ((read = in.read(buffer)) != -1) {
                digest.update(buffer, 0, read);
            }
        }
        byte[] hash = digest.digest();
        StringBuilder sb = new StringBuilder();
        for (byte b : hash) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }

    public static byte[] encryptAesGcmToCombined(byte[] plaintext, byte[] key, byte[] iv) throws Exception {
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        SecretKeySpec keySpec = new SecretKeySpec(key, "AES");
        GCMParameterSpec gcmSpec = new GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv);
        cipher.init(Cipher.ENCRYPT_MODE, keySpec, gcmSpec);
        return cipher.doFinal(plaintext);
    }

    public static byte[] decryptAesGcmFromCombined(byte[] combined, byte[] key, byte[] iv) throws Exception {
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        SecretKeySpec keySpec = new SecretKeySpec(key, "AES");
        GCMParameterSpec gcmSpec = new GCMParameterSpec(GCM_TAG_LENGTH_BITS, iv);
        cipher.init(Cipher.DECRYPT_MODE, keySpec, gcmSpec);
        return cipher.doFinal(combined);
    }

    public static ByteEncryptionResult encryptBytesDeterministic(byte[] plaintext, byte[] key, byte[] iv) throws Exception {
        byte[] combined = encryptAesGcmToCombined(plaintext, key, iv);
        int tagLength = 16;
        int ctLength = combined.length - tagLength;
        byte[] ciphertext = new byte[ctLength];
        byte[] tagBytes = new byte[tagLength];
        System.arraycopy(combined, 0, ciphertext, 0, ctLength);
        System.arraycopy(combined, ctLength, tagBytes, 0, tagLength);
        return new ByteEncryptionResult(ciphertext, iv, tagBytes);
    }

    public static class EncryptionResult {
        public final boolean success;
        public final byte[] iv;
        public final byte[] tag;
        public final Map<String, Object> fileInfo;

        public EncryptionResult(boolean success, byte[] iv, byte[] tag, Map<String, Object> fileInfo) {
            this.success = success;
            this.iv = iv;
            this.tag = tag;
            this.fileInfo = fileInfo;
        }
    }

    public static class ByteEncryptionResult {
        public final byte[] ciphertext;
        public final byte[] iv;
        public final byte[] tag;

        public ByteEncryptionResult(byte[] ciphertext, byte[] iv, byte[] tag) {
            this.ciphertext = ciphertext;
            this.iv = iv;
            this.tag = tag;
        }
    }
}
